analyzers/src/SonarAnalyzer.CSharp/Rules/UseUriInsteadOfString.cs

```diff
@@ -121,85 +121,74 @@ private static void VerifyPropertyDeclaration(SyntaxNodeAnalysisContext context)
-            {
-                return;
-            }
-
-            var stringUrlParams = GetStringUrlParamIndexes(invokedMethodSymbol);
-            var methodOverloads = FindOverloadsThatUseUriTypeInPlaceOfString(invokedMethodSymbol, stringUrlParams);
-            if (stringUrlParams.Count > 0 && methodOverloads.Any())
+            if (context.SemanticModel.GetSymbolInfo(context.Node).Symbol is IMethodSymbol invokedMethodSymbol && !invokedMethodSymbol.IsInType(KnownType.System_Uri))
             {
-                context.ReportIssue(Diagnostic.Create(RuleS4005, context.Node.GetLocation()));
+                var stringUrlParams = StringUrlParamIndexes(invokedMethodSymbol);
+                var methodOverloads = FindOverloadsThatUseUriTypeInPlaceOfString(invokedMethodSymbol, stringUrlParams);
+                if (stringUrlParams.Any() && methodOverloads.Any())
+                {
+                    context.ReportIssue(Diagnostic.Create(RuleS4005, context.Node.GetLocation()));
+                }
             }
         }
 
         private static void VerifyReturnType(SyntaxNodeAnalysisContext context, BaseMethodDeclarationSyntax methodDeclaration, IMethodSymbol methodSymbol)
         {
-            var returnTypeLocation = (methodDeclaration as MethodDeclarationSyntax)?.ReturnType?.GetLocation();
-
-            if (returnTypeLocation != null
+            if ((methodDeclaration as MethodDeclarationSyntax)?.ReturnType?.GetLocation() is { } returnTypeLocation
                 && methodSymbol.ReturnType.Is(KnownType.System_String)
                 && NameContainsUri(methodSymbol.Name))
             {
                 context.ReportIssue(Diagnostic.Create(RuleS3995, returnTypeLocation));
             }
         }
 
-        private static IEnumerable<IMethodSymbol> FindOverloadsThatUseUriTypeInPlaceOfString(IMethodSymbol originalMethodSymbol, ICollection<int> paramIdx)
+        private static IEnumerable<IMethodSymbol> FindOverloadsThatUseUriTypeInPlaceOfString(IMethodSymbol originalMethodSymbol, ISet<int> paramIdx)
         {
-            if (paramIdx.Count == 0)
+            if (paramIdx.Any())
             {
-                yield break;
-            }
-
-            foreach (var methodSymbol in GetOtherMethodOverrides(originalMethodSymbol))
-            {
-                if (methodSymbol.Parameters.Where((paramSymbol, i) => UsesUriInPlaceOfStringUri(paramSymbol, originalMethodSymbol.Parameters[i], paramIdx.Contains(i))).Any())
+                foreach (var methodSymbol in OtherMethodOverrides(originalMethodSymbol))
                 {
-                    yield return methodSymbol;
+                    if (methodSymbol.Parameters.Where((x, index) => UsesUriInPlaceOfStringUri(x, originalMethodSymbol.Parameters[index], paramIdx.Contains(index))).Any())
+                    {
+                        yield return methodSymbol;
+                    }
                 }
             }
```

Comment:
Can be turned into a LINQ query.

Reply:
I tried,but it was ugly (too much nesting)